dbms/src/Flash/Pipeline/Schedule/Events/Event.h

```diff
@@ -0,0 +1,99 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Common/MemoryTracker.h>
+#include <Flash/Executor/PipelineExecutorStatus.h>
+#include <Flash/Pipeline/Schedule/Tasks/Task.h>
+
+#include <atomic>
+#include <memory>
+#include <vector>
+
+namespace DB
+{
+enum class EventStatus
+{
+    INIT,
+    SCHEDULED,
+    FINISHED,
+};
+
+class Event;
+using EventPtr = std::shared_ptr<Event>;
+using Events = std::vector<EventPtr>;
+
+class Event : public std::enable_shared_from_this<Event>
+{
+public:
+    Event(PipelineExecutorStatus & exec_status_, MemoryTrackerPtr mem_tracker_)
+        : exec_status(exec_status_)
+        , mem_tracker(std::move(mem_tracker_))
+    {}
+    virtual ~Event() = default;
+
+    void addInput(const EventPtr & input);
+
+    // schedule, onTaskFinish and finish maybe called directly in TaskScheduler,
+    // so these functions must be noexcept.
+    void schedule() noexcept;
+
+    void onTaskFinish() noexcept;
+
+    bool withoutInput();
+
+    bool isCancelled()
+    {
+        return exec_status.isCancelled();
+    }
+
+    void toError(std::string && err_msg);
```

Comment:
I think maybe we should remove this method:
1. It only delegates `exec_status.toError()`;
2. It's all used internally in `Event`.

Reply:
ok, removed.